### ISSUE TITLE
Nondeterministic behavior in CountingGreyboxFuzzer

### ISSUE THREAD

**Issue by galois-agrushin** [uds-se/fuzzingbook#138]
**Describe the bug**

The `CountingGreyboxFuzzer` in the [Greybox Fuzzing](https://www.fuzzingbook.org/html/GreyboxFuzzer.html) chapter of the Fuzzing Book sometimes has nondeterministic behavior, i.e., generates a different sequence of inputs for the same target program, under the same random number generator seed.  This seems to stem from the use of the `getPathID` method with coverage passed as a set: the same logical set may be serialized differently, and thus map to different hash values, which are used as path IDs.

**To Reproduce**

The following code, if executed with a high enough number of trials, may print different inputs to the console for different executions:

```
import random
import time
from fuzzingbook import GreyboxFuzzer as gf
from fuzzingbook import Coverage as cov

random.seed(1)
    
maze_string = """
+-+-----+
|X|     |
| | --+ |
| |   | |
| +-- | |
|     |#|
+-----+-+
"""

maze_code = gf.generate_maze_code(maze_string)
exec( maze_code )

seed_input = " "
fast_schedule = gf.AFLFastSchedule(5)
fast_fuzzer = gf.CountingGreyboxFuzzer([seed_input], gf.MazeMutator(['L', 'R', 'U', 'D']), fast_schedule)
fast_fuzzer.runs(gf.FunctionCoverageRunner(maze), trials=50000)
print(fast_fuzzer.inputs[-1])
```

**Expected behavior**

The behavior of the fuzzer should be fully deterministic, so long as the random seed is fixed via `random.seed` before the fuzzer is used.  This can be achieved, for example, by sorting the set, prior to serializing it in `getPathID`, i.e., `pickled = pickle.dumps(sorted(coverage))`.

**Desktop (please complete the following information):**

- OS: Ubuntu 20.04.3 LTS
- Browser: Firefox, running JupyterLab
- Python version: 3.9.7

**Comment by mboehme** [comment]
Hi @galois-agrushin, thanks for spotting this! Sorting the coverage elements sounds reasonable. Can you submit a Pull Request?

**Comment by galois-agrushin** [comment]
Hi @mboehme; I ran some local tests, and the issue seems to be resolved with an adjusted version of `getPathID`:

```
def getPathID(coverage: Any) -> str:
    pickled = pickle.dumps(sorted(coverage))
    return hashlib.md5(pickled).hexdigest()
```

I searched through the repository, and the definition of `getPathID` seems to appear in the following files:

- docs/beta/code/GreyboxFuzzer.py
- docs/beta/html/GreyboxFuzzer.html
- docs/beta/notebooks/Greybox_Fuzzing.ipynb
- docs/beta/notebooks/GreyboxFuzzer.ipynb
- docs/beta/slides/GreyboxFuzzer.slides.html
- docs/code/GreyboxFuzzer.py
- docs/html/GreyboxFuzzer.html
- docs/notebooks/GreyboxFuzzer.ipynb
- docs/slides/GreyboxFuzzer.slides.html
- notebooks/GreyboxFuzzer.ipynb

Should the change only be made to `notebooks/GreyboxFuzzer.ipynb`, or to other files as well?

**Comment by mboehme** [comment]
Thanks! Only under notebooks/GreyboxFuzzer.ipynb should be sufficient. Everything else should be auto-generated in the next iteration (ping @vrthra: is that correct?)

**Comment by vrthra** [comment]
Yes, only `ipynb` files in `notebooks/` need to be modified. Everything else is auto-generated.

**Comment by galois-agrushin** [comment]
Thank you for the clarification; I created the [Pull Request](https://github.com/uds-se/fuzzingbook/pull/139).

**Comment by mboehme** [comment]
Closed via #139
